File: app/src/main/java/com/example/cameraaccess/data/model/AppError.kt
package com.example.cameraaccess.data.model

sealed class AppError : Exception() {
    abstract val userMessage: String

    data class NetworkError(val originalCause: Throwable? = null) : AppError() {
        override val userMessage: String = "Connection problem. Please check your internet and try again."
    }

    data class AuthError(override val message: String = "Invalid email or password.") : AppError() {
        override val userMessage: String = message
    }

    data class ServerError(val code: Int, override val message: String? = null) : AppError() {
        override val userMessage: String = "Server error ($code). Please try again later."
    }

    data class ValidationError(override val message: String) : AppError() {
        override val userMessage: String = message
    }

    data class UnexpectedError(val originalCause: Throwable? = null) : AppError() {
        override val userMessage: String = "An unexpected error occurred. Please try again."
    }
}

File: TauResearch/Models/AppError.swift
// AppError.swift
// Mirrors AppError.kt — typed errors with a user-facing message.

import Foundation

enum AppError: LocalizedError {
    case network(underlying: Error?)
    case auth(message: String = "Invalid email or password.")
    case server(code: Int, message: String? = nil)
    case validation(message: String)
    case unexpected(underlying: Error? = nil)

    var userMessage: String {
        switch self {
        case .network:
            return "Connection problem. Please check your internet and try again."
        case .auth(let m):
            return m
        case .server(let code, _):
            return "Server error (\(code)). Please try again later."
        case .validation(let m):
            return m
        case .unexpected:
            return "An unexpected error occurred. Please try again."
        }
    }

    var errorDescription: String? { userMessage }
}
